man/man1/doc2lt.1
.TH doc2lt 1 "6/25/2019" " " ""
.SH NAME
doc2lt \-  program to extract short descriptions in various formats 
.SH INPUT
.PD 0
.TP
.B -html     
- Use tabular format output in HTML
.PD 1
.PD 0
.TP
.B -mapref path 
- Use path as a 
.B map
file; this contains mappings from
names to links.  Useful mostly with 
.I -html
\&.
Multiple 
.I -mapref
files
may be given.
.PD 1
.PD 0
.TP
.B -defn path 
- Use 
.I path
to redefine the output commands.  See the default
definition files 
.I latex.def
and 
.I html.def
for examples.
.PD 1
.PD 0
.TP
.B filenames 
- Names the files from which documents are to be extracted
.PD 1

.SH NOTES
The default format is LaTeX.

The alternate definition file 
.I htmltable.def
creates a Netscape-style table
for HTML.

.SH AUTHOR
William Gropp
.SH LOCATION
doc2lt.cc


RELEASE_NOTES
Release Notes for the Sowing package

1.1.26: June 23, 2019

 Added these release notes.
 This version is an update from 1.1.25.  The major updates are:
 doctext: Added an option to handled a single ". " command with the same
   formatting as a list created with :+ ... -". Also allow "- " to define
   a list of a single element.  This is to improve the formatting of manual
   pages. The argument -oldstylearg will tell doctext to use the approach
   from previous versions. With the new style of argument command, doctext
   will issue a warning message for argument lists of more than one argument
   that do not start with the "+" command.  The the -Wargdesc option,
   doctext will also warn about arguments that do not include a description;
   these should use a different formatting choice.
   doctext environment variables:
      DOCTEXT_VERBOSE if set to yes or true will turn on verbose output
      DOCTEXT_OLDSTYLEARG if set to yes or true will use the processing of
      the "+", ". ", and "-" commands from previous versions of doctext
 bfort: Added -o filename for the output filename. Use const to avoid CtoF
   transformations (const data can't be changed in the called routine). Improved
   error and warning messages. Better formatting for output interface files.
   Better support for converting from and to Fortran strings, with options
      -fstring      - Generate code to convert to/from Fortran character
                      strings. Only lightly tested so far
      -fnostring    - Do not generate interface files for routines with char*
                      arguments
      -fstring-asis - Close to the behavior in previous versions - just
                      pass the pointer to the string from Fortran to the
                      C routine.  This is usually incorrect, however.
   The environment variable BFORT_STRINGHANDLING can be set to IGNORE to
   get the effect of -fstring-asis.
 bfort tests: There is a new and cleaner set of tests in testing2 ; this
   is intended to become the new test set in the next release or so.
 bfort2: New version of bfort that uses two passes to improve the readability
   of the generated interfaces.  Build with make BFORTVERSION=new ; test in
   bfort/testing2 with make BFORTVERSION=new testing . This version will
   eventually replace bfort.
 configure: Added an update to the enable-strict rules, using more recent
   autoconf macros from MPICH.  Added C++ to the strict rules
 make dist: Changed the make dist rules to explicitly select files, rather
   than taking the current svn commited files (which led to the distributions
   using different files in confdb than in the version used for development)
   general code: Mostly clean compile with --enable-strict, except for tr.c ,
   which is the memory allocator.
 misc: Some cleanup of files, particularly space squashes. Preparations to
   move from svn to git.

1.1.25: September 27, 2016
 bfort: partial update to remove support for non-ANSI-C prototypes, and a
   partial add of support for transforming between C and Fortran character
   strings
 Misc: other changes and cleanup

src/doctext/doc2lt.cc
#include "textout.h"
#include "cmdline.h"
#include "docutil.h"
#include "docpath.h"
#include "inutil.h"
#include "maptok.h"

int DocOutRoutine( InStream *ins, TextOut *text, char *matchstring );
int DocOutMacro( InStream *ins, TextOut *text, char *matchstring );

char *IgnoreString = 0;

int verbose = 0;

char NewlineString[3];
/*D
doc2lt - program to extract short descriptions in various formats

Input:
+ -html     - Use tabular format output in HTML
. -mapref path - Use path as a `map` file; this contains mappings from
  names to links.  Useful mostly with '-html'.  Multiple '-mapref' files
  may be given.
. -defn path - Use 'path' to redefine the output commands.  See the default
   definition files 'latex.def' and 'html.def' for examples.
- filenames - Names the files from which documents are to be extracted

Notes:
The default format is LaTeX.

The alternate definition file 'htmltable.def' creates a Netscape-style table
for HTML.

Author:
William Gropp
D*/

int main( int argc, char ** argv )
{
    const char   *filename=0;
    InStream     *ins, *incommands, *mapins;
    OutStream    *outs;
    OutStreamMap *map;
    TextOut      *textout;
    CmdLine      *cmd;
    char         matchstring[5];
    const char   *path;

    cmd = new CmdLine( argc, argv );

    // Output to stdout
    outs       = new OutStreamFile();

    // Check for mapping files; there should be a routine to do this
    // Also, should do this ONLY for -html form (Map streams currently
    // only for HTML).
    map = 0;
    while (!cmd->GetArgPtr( "-mapref", &path )) {
    	if (!map)
	    map    = new OutStreamMap( outs, 128 );
	mapins = new InStreamFile( path, "r" );
	if (mapins->status) {
	    fprintf( stderr, "Could not read map file %s\n", path );
	    }
	else
	    map->ReadMap( mapins, 0 );
	delete mapins;
	}
    if (map)
	outs = map;

    // Check for output format
    if (!cmd->HasArg( "-html" )) {
	textout    = new TextOutHTML( outs );
	incommands = new InStreamFile(
		      DOCTEXT_PATH, "DOCTEXT_PATH", "html.def", "r" );
	}
    else {
	textout    = new TextOutTeX( outs );
	incommands = new InStreamFile(
		      DOCTEXT_PATH, "DOCTEXT_PATH", "latex.def", "r" );
	}
    textout->SetDebug( 0 );
    textout->ReadCommands( incommands );
    incommands->Close();

    // Allow the user to define replacement commands
    while (!cmd->GetArgPtr( "-defn", &path )) {
	incommands = new InStreamFile( path, "r" );
	if (incommands->status) {
	    fprintf( stderr, "Could not open definition file %s\n", path );
	    }
	else
	    textout->ReadCommands( incommands );
	incommands->Close();
	}

    textout->PutOp( "preamble" );
    while (!cmd->NextArg( &filename )) {
	ins = new InStreamFile( filename, "r" );
	if (ins->status) {
	    fprintf( stderr, "Could not open file %s\n", filename );
	    continue;
	    }
	while (!FindPattern( ins, "/*[@M]", matchstring )) {
	    // First, need to remove any sub-options (C,X)
	    DocGetSubOptions( ins );
	    if (matchstring[2] == '@') {
		/* C routine */
		DocOutRoutine( ins, textout, matchstring );
		}
	    else {
		/* Macro (M) */
		DocOutMacro( ins, textout, matchstring );
		}
	    }
	delete ins;
	}
    textout->PutOp( "postamble" );
    delete outs;
    return 0;
}

/*
   To process the structured comment for this code, we need to
   (1) Get the description
   (2) Get the synopsis

   The description is of the form
   [space]*<name>[space]*[-]*[text,including\n]\n\n
   That is, two consequtive newlines end the description.

   The synopsis depends on the form.  For routines, it is read by
   reaching the end of the comment and then reading until the first {.
   For macros, we search for a "Synopsis:" line, then read until
   two newlines.
 */

/*
   The actual format, for LaTeX, is something like

   \k{routinename}{\CoDe{synopsis}\DeFn{definition}

   An HTML version might be

   <TR><TD>routinename</TD><TD>definition</TD></TR>

   with no synopsis.

   We can do this by using memory buffer outstreams, and then
   calling the output routines.
 */
int DocOutRoutine( InStream *ins, TextOut *text, char *matchstring )
{
    char routinename[128];
    int  at_end;
    OutStreamBuf *outdesc, *outsynop;
    TextOutStrm  *textout, *toutsynop;

    outdesc = new OutStreamBuf( 1024 );
    textout = new TextOutStrm( outdesc );

    if (DocReadName( ins, routinename, 128 )) return 1;
    if (DocReadDescription( ins, matchstring, textout, 0, &at_end )) return 1;
    if (!at_end)
      if (DocSkipToFuncSynopsis( ins, matchstring )) return 1;

    outsynop = new OutStreamBuf( 16000 );
    toutsynop = new TextOutStrm(outsynop);
    if (DocReadFuncSynopsis( ins, toutsynop )) return 1;

    // Generate the output.
    text->PutOp( "key", routinename );
    // Horrible hack which indicates a design flaw.  Will need to fix
    // eventually
    text->PutOp( "synopsis", ((OutStreamBuf *)(toutsynop->out))->GetBuffer() );
    text->PutOp( "definition", outdesc->GetBuffer() );

    delete outdesc;
    delete outsynop;
    delete textout;
    delete toutsynop;

    return 0;
}

int DocOutMacro( InStream *ins, TextOut *text, char *matchstring )
{
    char routinename[128];
    OutStreamBuf *outdesc, *outsynop;
    TextOutStrm  *textout;
    int at_end;

    outdesc = new OutStreamBuf( 1024 );
    textout = new TextOutStrm( outdesc );

    if (DocReadName( ins, routinename, 128 )) return 1;
    if (DocReadDescription( ins, matchstring, textout, 0, &at_end )) return 1;
    if (at_end) {
        fprintf( stderr, "No synopsis for macro %s\n", routinename );
	return 1;
        }
    if (DocSkipToMacroSynopsis( ins, matchstring )) return 1;

    outsynop = new OutStreamBuf( 16000 );
    if (DocReadMacroSynopsis( ins, matchstring, textout/*outsynop*/, &at_end )) return 1;

    // Generate the output.
    text->PutOp( "key", routinename );
    text->PutOp( "synopsis", outsynop->GetBuffer() );
    text->PutOp( "definition", outdesc->GetBuffer() );
    delete outdesc;
    delete outsynop;
    delete textout;

    return 0;
}

// These are needed for the Enum code, but only for the index case, which
// is not supported for doc2lt
const char *GetCurrentRoutinename( void )
{
    return "";
}

const char *GetCurrentInputFileName(void)
{
    return "";
}

const char *GetCurrentFileName( void )
{
    return "";
}

